File: android/src/main/kotlin/com/ahmtydn/calendar_bridge/CalendarBridgePlugin.kt
package com.ahmtydn.calendar_bridge

import android.app.Activity
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** CalendarBridgePlugin */
class CalendarBridgePlugin: FlutterPlugin, MethodCallHandler, ActivityAware, 
    PluginRegistry.RequestPermissionsResultListener {
  private lateinit var channel: MethodChannel
  private lateinit var calendarManager: CalendarManager
  private lateinit var eventManager: EventManager
  private var activity: Activity? = null
  private val scope = CoroutineScope(Dispatchers.Main)
  private var permissionResult: Result? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.ahmtydn/calendar_bridge")
    channel.setMethodCallHandler(this)
    
    calendarManager = CalendarManager(flutterPluginBinding.applicationContext)
    eventManager = EventManager(flutterPluginBinding.applicationContext)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    scope.launch {
      try {
        when (call.method) {
          "requestPermissions" -> {
            handleRequestPermissions(result)
          }
          
          "hasPermissions" -> {
            val hasPermission = calendarManager.hasPermissions()
            result.success(hasPermission)
          }
          
          "retrieveCalendars" -> {
            handleRetrieveCalendars(result)
          }
          
          "retrieveEvents" -> {
            handleRetrieveEvents(call, result)
          }
          
          "createEvent" -> {
            handleCreateEvent(call, result)
          }
          
          "updateEvent" -> {
            handleUpdateEvent(call, result)
          }
          
          "deleteEvent" -> {
            handleDeleteEvent(call, result)
          }
          
          "createCalendar" -> {
            handleCreateCalendar(call, result)
          }
          
          "deleteCalendar" -> {
            handleDeleteCalendar(call, result)
          }
          
          "retrieveCalendarColors" -> {
            handleRetrieveCalendarColors(result)
          }
          
          "retrieveEventColors" -> {
            handleRetrieveEventColors(call, result)
          }
          
          "updateCalendarColor" -> {
            handleUpdateCalendarColor(call, result)
          }
          
          "deleteEventInstance" -> {
            handleDeleteEventInstance(call, result)
          }
          
          else -> {
            result.notImplemented()
          }
        }
      } catch (e: Exception) {
        handleError(e, result)
      }
    }
  }

  private fun handleRequestPermissions(result: Result) {
    if (calendarManager.hasPermissionsBoolean()) {
      result.success(true)
      return
    }
    val currentActivity = activity
    if (currentActivity == null) {
      result.success(false)
      return
    }
    permissionResult = result
    calendarManager.requestPermissions(currentActivity)
  }

  override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<out String>,
    grantResults: IntArray
  ): Boolean {
    if (requestCode == CalendarManager.PERMISSION_REQUEST_CODE) {
      val granted = calendarManager.hasPermissionsBoolean()
      permissionResult?.let {
        it.success(granted)
        permissionResult = null
      }
      return true
    }
    return false
  }

  private suspend fun handleRetrieveCalendars(result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val calendars = calendarManager.retrieveCalendars()
    result.success(calendars)
  }

  private suspend fun handleRetrieveEvents(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val calendarId = arguments["calendarId"] as? String
      ?: throw CalendarException.InvalidArgument("Calendar ID is required")
    
    val events = eventManager.retrieveEvents(calendarId, arguments)
    result.success(events)
  }

  private suspend fun handleCreateEvent(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val eventId = eventManager.createEvent(arguments)
    result.success(eventId)
  }

  private suspend fun handleUpdateEvent(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val eventId = eventManager.updateEvent(arguments)
    result.success(eventId)
  }

  private suspend fun handleDeleteEvent(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val calendarId = arguments["calendarId"] as? String
      ?: throw CalendarException.InvalidArgument("Calendar ID is required")
    
    val eventId = arguments["eventId"] as? String
      ?: throw CalendarException.InvalidArgument("Event ID is required")
    
    val success = eventManager.deleteEvent(calendarId, eventId)
    result.success(success)
  }

  private suspend fun handleCreateCalendar(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val calendar = calendarManager.createCalendar(arguments)
    result.success(calendar)
  }

  private suspend fun handleDeleteCalendar(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val calendarId = arguments["calendarId"] as? String
      ?: throw CalendarException.InvalidArgument("Calendar ID is required")
    
    val success = calendarManager.deleteCalendar(calendarId)
    result.success(success)
  }

  private fun handleError(error: Throwable, result: Result) {
    when (error) {
      is CalendarException -> {
        result.error(error.code, error.message, error.details)
      }
      else -> {
        result.error("UNKNOWN_ERROR", "An unexpected error occurred", error.message)
      }
    }
  }

  private suspend fun handleRetrieveCalendarColors(result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val colors = calendarManager.getCalendarColors()
    result.success(colors)
  }

  private suspend fun handleRetrieveEventColors(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val calendarId = arguments["calendarId"] as? String
      ?: throw CalendarException.InvalidArgument("Missing calendarId")
    
    val colors = calendarManager.getEventColors(calendarId)
    result.success(colors)
  }

  private suspend fun handleUpdateCalendarColor(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val calendarId = arguments["calendarId"] as? String
      ?: throw CalendarException.InvalidArgument("Missing calendarId")
      
    val colorKey = arguments["colorKey"] as? String
      ?: throw CalendarException.InvalidArgument("Missing colorKey")
    
    val success = calendarManager.updateCalendarColor(calendarId, colorKey)
    result.success(success)
  }

  private suspend fun handleDeleteEventInstance(call: MethodCall, result: Result) {
    if (!calendarManager.hasPermissionsBoolean()) {
      throw CalendarException.PermissionDenied()
    }
    
    val arguments = call.arguments as? Map<String, Any>
      ?: throw CalendarException.InvalidArgument("Missing arguments")
    
    val calendarId = arguments["calendarId"] as? String
      ?: throw CalendarException.InvalidArgument("Missing calendarId")
      
    val eventId = arguments["eventId"] as? String
      ?: throw CalendarException.InvalidArgument("Missing eventId")
      
    val startDate = arguments["startDate"] as? Long
      ?: throw CalendarException.InvalidArgument("Missing startDate")
      
    val followingInstances = arguments["followingInstances"] as? Boolean ?: false
    
    val success = eventManager.deleteEventInstance(calendarId, eventId, startDate, followingInstances)
    result.success(success)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
    calendarManager.setActivity(activity)
    binding.addRequestPermissionsResultListener(this)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    activity = null
    calendarManager.setActivity(null)
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
    calendarManager.setActivity(activity)
    binding.addRequestPermissionsResultListener(this)
  }

  override fun onDetachedFromActivity() {
    activity = null
    calendarManager.setActivity(null)
  }
}

File: darwin/calendar_bridge/Sources/calendar_bridge/CalendarBridgePlugin.swift
#if os(iOS)
import Flutter
import UIKit
#elseif os(macOS)
import Cocoa
import FlutterMacOS
#endif
import EventKit
import Foundation

public class CalendarBridgePlugin: NSObject, FlutterPlugin {
    private let eventStore = EKEventStore()
    private let calendarManager: CalendarManager
    private let eventManager: EventManager
    
    override init() {
        self.calendarManager = CalendarManager(eventStore: eventStore)
        self.eventManager = EventManager(eventStore: eventStore)
        super.init()
    }
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        #if os(iOS)
        let channel = FlutterMethodChannel(
            name: "com.ahmtydn/calendar_bridge",
            binaryMessenger: registrar.messenger()
        )
        #elseif os(macOS)
        let channel = FlutterMethodChannel(
            name: "com.ahmtydn/calendar_bridge",
            binaryMessenger: registrar.messenger
        )
        #endif
        let instance = CalendarBridgePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            do {
                switch call.method {
                case "requestPermissions":
                    let granted = await requestPermissions()
                    result(granted)
                    
                case "hasPermissions":
                    let permissionStatus = hasPermissions()
                    result(permissionStatus)
                    
                case "retrieveCalendars":
                    try await handleRetrieveCalendars(result: result)
                    
                case "retrieveEvents":
                    try await handleRetrieveEvents(call: call, result: result)
                    
                case "createEvent":
                    try await handleCreateEvent(call: call, result: result)
                    
                case "updateEvent":
                    try await handleUpdateEvent(call: call, result: result)
                    
                case "deleteEvent":
                    try await handleDeleteEvent(call: call, result: result)
                    
                case "createCalendar":
                    try await handleCreateCalendar(call: call, result: result)
                    
                case "deleteCalendar":
                    try await handleDeleteCalendar(call: call, result: result)
                    
                case "retrieveCalendarColors":
                    try await handleRetrieveCalendarColors(result: result)
                    
                case "retrieveEventColors":
                    try await handleRetrieveEventColors(call: call, result: result)
                    
                case "updateCalendarColor":
                    try await handleUpdateCalendarColor(call: call, result: result)
                    
                case "deleteEventInstance":
                    try await handleDeleteEventInstance(call: call, result: result)
                    
                default:
                    result(FlutterMethodNotImplemented)
                }
            } catch {
                handleError(error: error, result: result)
            }
        }
    }
    
    // MARK: - Permission Management
    
    private func requestPermissions() async -> Bool {
        #if os(iOS)
        if #available(iOS 17.0, *) {
            do {
                return try await eventStore.requestFullAccessToEvents()
            } catch {
                return false
            }
        } else {
            return await withCheckedContinuation { continuation in
                eventStore.requestAccess(to: .event) { granted, _ in
                    continuation.resume(returning: granted)
                }
            }
        }
        #elseif os(macOS)
        return await withCheckedContinuation { continuation in
            eventStore.requestAccess(to: .event) { granted, error in              
                continuation.resume(returning: granted)
            }
        }
        #endif
    }
    
    private func hasPermissions() -> String {
        let status: EKAuthorizationStatus
        #if os(iOS)
        if #available(iOS 17.0, *) {
            status = EKEventStore.authorizationStatus(for: .event)
            switch status {
            case .fullAccess:
                return "granted"
            case .writeOnly:
                return "restricted"
            case .denied:
                return "denied"
            case .notDetermined:
                return "notDetermined"
            @unknown default:
                return "notDetermined"
            }
        } else {
            status = EKEventStore.authorizationStatus(for: .event)
            switch status {
            case .authorized:
                return "granted"
            case .denied:
                return "denied"
            case .restricted:
                return "restricted"
            case .notDetermined:
                return "notDetermined"
            @unknown default:
                return "notDetermined"
            }
        }
        #elseif os(macOS)
        status = EKEventStore.authorizationStatus(for: .event)
        switch status {
        case .authorized:
            return "granted"
        case .denied:
            return "denied"
        case .restricted:
            return "restricted"
        case .notDetermined:
            return "notDetermined"
        case .writeOnly:
            return "restricted" 
        case .fullAccess:
            return "granted"
        @unknown default:
            return "notDetermined"
        }
        #endif
    }
    
    private func checkPermissions() throws {
        guard hasPermissions() == "granted" else {
            throw CalendarError.permissionDenied
        }
    }
    
    // MARK: - Calendar Operations
    
    private func handleRetrieveCalendars(result: @escaping FlutterResult) async throws {
        try checkPermissions()
        let calendars = try calendarManager.retrieveCalendars()
        result(calendars)
    }
    
    private func handleCreateCalendar(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        try checkPermissions()
        guard let arguments = call.arguments as? [String: Any] else {
            throw CalendarError.invalidArgument("Missing arguments")
        }
        let calendar = try calendarManager.createCalendar(arguments: arguments)
        result(calendar)
    }
    
    private func handleDeleteCalendar(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        try checkPermissions()
        guard let arguments = call.arguments as? [String: Any],
              let calendarId = arguments["calendarId"] as? String else {
            throw CalendarError.invalidArgument("Calendar ID is required")
        }
        let success = try calendarManager.deleteCalendar(calendarId: calendarId)
        result(success)
    }
    
    // MARK: - Event Operations
    
    private func handleRetrieveEvents(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        try checkPermissions()
        guard let arguments = call.arguments as? [String: Any],
              let calendarId = arguments["calendarId"] as? String else {
            throw CalendarError.invalidArgument("Calendar ID is required")
        }
        let events = try eventManager.retrieveEvents(calendarId: calendarId, arguments: arguments)
        result(events)
    }
    
    private func handleCreateEvent(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        try checkPermissions()
        guard let arguments = call.arguments as? [String: Any] else {
            throw CalendarError.invalidArgument("Missing arguments")
        }
        let eventId = try eventManager.createEvent(arguments: arguments)
        result(eventId)
    }
    
    private func handleUpdateEvent(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        try checkPermissions()
        guard let arguments = call.arguments as? [String: Any] else {
            throw CalendarError.invalidArgument("Missing arguments")
        }
        let eventId = try eventManager.updateEvent(arguments: arguments)
        result(eventId)
    }
    
    private func handleDeleteEvent(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        try checkPermissions()
        guard let arguments = call.arguments as? [String: Any],
              let calendarId = arguments["calendarId"] as? String,
              let eventId = arguments["eventId"] as? String else {
            throw CalendarError.invalidArgument("Calendar ID and Event ID are required")
        }
        let success = try eventManager.deleteEvent(calendarId: calendarId, eventId: eventId)
        result(success)
    }
    
    // MARK: - Color Management
    
    private func handleRetrieveCalendarColors(result: @escaping FlutterResult) async throws {
        // iOS/macOS doesn't have predefined calendar colors like Android
        // We return some common colors that work well
        let colors: [String: Int] = [
            "red": 0xFFFF0000,
            "green": 0xFF00FF00,
            "blue": 0xFF0000FF,
            "yellow": 0xFFFFFF00,
            "orange": 0xFFFFA500,
            "purple": 0xFF800080,
            "pink": 0xFFFFC0CB,
            "brown": 0xFFA52A2A,
            "gray": 0xFF808080,
            "black": 0xFF000000
        ]
        result(colors)
    }
    
    private func handleRetrieveEventColors(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        // iOS/macOS doesn't have calendar-specific event colors
        // Return the same common colors
        let colors: [String: Int] = [
            "red": 0xFFFF0000,
            "green": 0xFF00FF00,
            "blue": 0xFF0000FF,
            "yellow": 0xFFFFFF00,
            "orange": 0xFFFFA500,
            "purple": 0xFF800080,
            "pink": 0xFFFFC0CB,
            "brown": 0xFFA52A2A,
            "gray": 0xFF808080,
            "black": 0xFF000000
        ]
        result(colors)
    }
    
    private func handleUpdateCalendarColor(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        try checkPermissions()
        
        guard let arguments = call.arguments as? [String: Any],
              let calendarId = arguments["calendarId"] as? String,
              let colorKey = arguments["colorKey"] as? String else {
            throw CalendarError.invalidArgument("Missing required arguments")
        }
        
        let success = try calendarManager.updateCalendarColor(calendarId: calendarId, colorKey: colorKey)
        result(success)
    }
    
    private func handleDeleteEventInstance(call: FlutterMethodCall, result: @escaping FlutterResult) async throws {
        try checkPermissions()
        
        guard let arguments = call.arguments as? [String: Any],
              let calendarId = arguments["calendarId"] as? String,
              let eventId = arguments["eventId"] as? String,
              let startDateMs = arguments["startDate"] as? Int64 else {
            throw CalendarError.invalidArgument("Missing required arguments")
        }
        
        let startDate = Date(timeIntervalSince1970: Double(startDateMs) / 1000.0)
        let followingInstances = arguments["followingInstances"] as? Bool ?? false
        
        let success = try await eventManager.deleteEventInstance(
            calendarId: calendarId,
            eventId: eventId,
            startDate: startDate,
            followingInstances: followingInstances
        )
        result(success)
    }
    
    // MARK: - Error Handling
    
    private func handleError(error: Error, result: @escaping FlutterResult) {
        let flutterError: FlutterError
        
        if let calendarError = error as? CalendarError {
            flutterError = calendarError.toFlutterError()
        } else {
            flutterError = FlutterError(
                code: "UNKNOWN_ERROR",
                message: "An unexpected error occurred",
                details: error.localizedDescription
            )
        }
        
        result(flutterError)
    }
}

// MARK: - Error Types

enum CalendarError: Error {
    case permissionDenied
    case calendarNotFound(String)
    case eventNotFound(String)  
    case invalidArgument(String)
    case unsupportedOperation(String)
    case platformError(String)
    
    func toFlutterError() -> FlutterError {
        switch self {
        case .permissionDenied:
            return FlutterError(
                code: "PERMISSION_DENIED",
                message: "Calendar permissions not granted",
                details: nil
            )
        case .calendarNotFound(let id):
            return FlutterError(
                code: "CALENDAR_NOT_FOUND",
                message: "Calendar not found",
                details: id
            )
        case .eventNotFound(let id):
            return FlutterError(
                code: "EVENT_NOT_FOUND", 
                message: "Event not found",
                details: id
            )
        case .invalidArgument(let message):
            return FlutterError(
                code: "INVALID_ARGUMENT",
                message: message,
                details: nil
            )
        case .unsupportedOperation(let operation):
            return FlutterError(
                code: "UNSUPPORTED_OPERATION",
                message: operation,
                details: nil
            )
        case .platformError(let message):
            return FlutterError(
                code: "PLATFORM_ERROR",
                message: message,
                details: nil
            )
        }
    }
}